scripts/asset_processor/video_asset_processor.py

```diff
@@ -3,529 +3,651 @@
+			self.do_process = False
+
+	def capture_to_array(self, capture):
+		"""
+		Function to convert OpenCV video capture to a list of
+		numpy arrays for faster processing and analysis.
+		@rtype: Tuple[list, np.ndarray, np.ndarray, int, int, int]
+		@param capture:
+		@return:  A tuple:
+					- list mapping indexes in returned sample frames to corresponding master samples
+					- sample frames
+					- sample frames HD
+					- total sample pixels count
+					- original sample height
+					- original sample width
+		@param q - quantization parameter to sample frames aligned by timestamp from source video and renditions
+		"""
+		# Create list of random timestamps in video file to calculate metrics at
+		if self.markup_master_frames:
+			self.master_indexes = np.sort(np.random.choice(self.total_frames, self.max_samples, False))
+			self.master_timestamps = []
+
+		# difference between master timestamp and best matching frame timestamp of current video
+		master_timestamp_diffs = [np.inf] * len(self.master_indexes)
+		# currently selected frames
+		candidate_frames = [None] * len(self.master_indexes)
+		# maps selected rendition sample to master sample
+		debug_index_mapping = {}
+		master_idx_map = []
+		frame_list = []
+		frame_list_hd = []
+		frames_read = 0
+		pixels = 0
+		height = 0
+		width = 0
+		timestamps_selected = []
+		# Iterate through each frame in the video
+		while True:
+			# Read the frame from the capture
+			frame_data = capture.read()
+			if frame_data is not None:
+				frames_read += 1
+				if self.markup_master_frames:
+					if frame_data.index in self.master_indexes:
+						self.master_timestamps.append(frame_data.timestamp)
+					else:
+						continue
+				# update candidate frames
+				ts_diffs = [abs(frame_data.timestamp - mts) for mts in self.master_timestamps]
+				best_match_idx = int(np.argmin(ts_diffs))
+				best_match = np.min(ts_diffs)
+				# max theoretical timestamp difference between 'matching' frames would be 1/(2*fps) + max(jitter)
+				# don't consider frames that are too far, otherwise the algorithm will be linear on memory vs video length
+				if best_match < 1 / (2 * capture.fps) and master_timestamp_diffs[best_match_idx] > best_match:
+					master_timestamp_diffs[best_match_idx] = best_match
+					candidate_frames[best_match_idx] = frame_data
+			# Break the loop when frames cannot be taken from original
+			else:
+				break
+
+		# process picked frames
+		for i in range(len(candidate_frames)):
+			frame_data = candidate_frames[i]
+			ts_diff = master_timestamp_diffs[i]
+			if frame_data is None or ts_diff > 1 / (2 * self.fps):
+				# no good matching candidate frame
+				continue
+			if self.debug_frames:
+				cv2.imwrite(f'{self.frame_dir_name}/{i:04}_{"m" if self.markup_master_frames else ""}_{frame_data.index}_{frame_data.timestamp:.4}.png', self._convert_debug_frame(frame_data.frame))
+			timestamps_selected.append(frame_data.timestamp)
+			master_idx_map.append(i)
```

Comment:
IIUC `master_idx_map` is supposed to map the index of a rendition frame to the corresponding index of a master frame. In that case, shouldn't this be `master_idx_map.append(self.master_indexes[i]`?

Reply:
This list is intended to keep correspondence between master and rendition frame lists in case some master frames are not matched. It's not linked with absolute frame index, because candidate frames are already selected at this point.

Reply:
Ah I see. Thanks for the clarification, I misunderstood how `master_idx_map` was being used.

Just to make sure I understand, suppose there are 5 master frame samples and the 2nd master frame is not matched for a particular rendition. Then, we would end up with the following:

- `master_idx_map = [0, 2, 3, 4]`
- A candidate frame list is generated for the rendition `frame_list = [candidate_frame_0, candidate_frame_1, candidate_frame_2, candidate_frame]`
- In `compare_renditions_instant()`:
    - `master_samples[0]` is the reference for `candidate_frame_0`
    - `master_samples[2]` is the reference for `candidate_frame_1`
    - `master_samples[3]` is the reference for `candidate_frame_2`
    - `master_samples[4]` is the reference for `candidate_frame_3`